external/storm-kafka-client/src/test/java/org/apache/storm/kafka/spout/MaxUncommittedOffsetTest.java

```diff
@@ -0,0 +1,245 @@
+import org.mockito.ArgumentCaptor;
+import org.mockito.MockitoAnnotations;
+
+public class MaxUncommittedOffsetTest {
+
+    @Rule
+    public KafkaUnitRule kafkaUnitRule = new KafkaUnitRule();
+
+    private final TopologyContext topologyContext = mock(TopologyContext.class);
+    private final Map<String, Object> conf = new HashMap<>();
+    private final SpoutOutputCollector collector = mock(SpoutOutputCollector.class);
+    private final long commitOffsetPeriodMs = 2_000;
+    private final int numMessages = 100;
+    private final int maxUncommittedOffsets = 10;
+    //This is set to be the same as maxUncommittedOffsets since it is difficult to test maxUncommittedOffsets when maxPollRecords is not the same
+    //If maxPollRecords is larger, a single call to poll will emit more than maxUncommittedOffsets messages
+    //If maxPollRecords is lower, it will cap how far past the commit offset the spout can read when there are failed tuples ready for retry
+    private final int maxPollRecords = maxUncommittedOffsets;
+    private final int initialRetryDelaySecs = 60;
+    private final KafkaSpoutConfig spoutConfig = getKafkaSpoutConfigBuilder(kafkaUnitRule.getKafkaPort())
+        .setOffsetCommitPeriodMs(commitOffsetPeriodMs)
+        .setMaxPollRecords(maxPollRecords)
+        .setMaxUncommittedOffsets(maxUncommittedOffsets)
+        .setRetry(new KafkaSpoutRetryExponentialBackoff(KafkaSpoutRetryExponentialBackoff.TimeInterval.seconds(initialRetryDelaySecs), KafkaSpoutRetryExponentialBackoff.TimeInterval.seconds(0),
+            1, KafkaSpoutRetryExponentialBackoff.TimeInterval.seconds(initialRetryDelaySecs))) //Retry once after a minute
+        .build();
+    private KafkaConsumer<String, String> consumerSpy;
+    private KafkaConsumerFactory<String, String> consumerFactory;
+    private KafkaSpout<String, String> spout;
+
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+        this.consumerSpy = spy(new KafkaConsumerFactoryDefault().createConsumer(spoutConfig));
+        this.consumerFactory = (kafkaSpoutConfig) -> consumerSpy;
+        this.spout = new KafkaSpout<>(spoutConfig, consumerFactory);
+    }
+
+    private void populateTopicData(String topicName, int msgCount) {
+        kafkaUnitRule.getKafkaUnit().createTopic(topicName);
+
+        IntStream.range(0, msgCount).forEach(value -> {
+            KeyedMessage<String, String> keyedMessage = new KeyedMessage<>(
+                topicName, Integer.toString(value),
+                Integer.toString(value));
+
+            kafkaUnitRule.getKafkaUnit().sendMessages(keyedMessage);
+        });
+    }
+
+    private void initializeSpout(int msgCount) {
+        populateTopicData(SingleTopicKafkaSpoutConfiguration.TOPIC, msgCount);
+        spout.open(conf, topologyContext, collector);
+        spout.activate();
+    }
+
+    public ArgumentCaptor<KafkaSpoutMessageId> emitMaxUncommittedOffsetsMessagesAndCheckNoMoreAreEmitted(int messageCount) {
```

Comment:
Minor: Might be better to include assertion for `messageCount >= maxUncommittedOffsets`.